### PR TITLE
replace scipy

### PR DESCRIPTION
**Pull request by hkctkuy**
Replace scipy clustering with rust according #62 issue

### CODE REVIEW

**Comment by SweetVishnya** [comment]
Closes #62

File: libcasr/src/stacktrace.rs [hunk 1]
```diff
@@ -190,67 +190,68 @@ pub fn dedup_stacktraces(stacktraces: &[Stacktrace]) -> Vec<bool> {
...
-        .spawn() else {
-        return Err(Error::Casr("Failed to launch python3".to_string()));
-    };
-    {
-        let python_stdin = python.stdin.as_mut().unwrap();
-        if python_stdin.write_all(lines.join("").as_bytes()).is_err() {
-            return Err(Error::Casr(
-                "Error while writing to stdin of python script".to_string(),
-            ));
+    let size = len * (len - 1) / 2;
```
Comment (anfedotoff) [inline libcasr/src/stacktrace.rs:197]: I found this code snippet from [here](https://docs.rs/kodama/latest/kodama/index.html#example)
```rust
// Build our condensed matrix by computing the dissimilarity between all
// possible coordinate pairs.
let mut condensed = vec![];
for row in 0..coordinates.len() - 1 {
    for col in row + 1..coordinates.len() {
        condensed.push(haversine(coordinates[row], coordinates[col]));
    }
}
```
I suppose we can do the same way using push?

File: libcasr/src/stacktrace.rs [hunk 2]
```diff
@@ -190,67 +190,68 @@ pub fn dedup_stacktraces(stacktraces: &[Stacktrace]) -> Vec<bool> {
...
+        }
     }
-    let output = String::from_utf8_lossy(&python.stdout);
-    let clusters = output
-        .split(' ')
-        .filter_map(|x| x.trim().parse::<u32>().ok())
-        .collect::<Vec<u32>>();
 
-    if clusters.len() != len {
+    if flat_clusters.len() != len {
```
Comment (Avgor46) [inline libcasr/src/stacktrace.rs:246]: Уже не нужно проверять

File: README.md [hunk 3]
```diff
@@ -74,8 +74,7 @@ $ cargo build --release
 4. Install runtime dependencies:
 
 ```
-$ sudo apt install gdb python3 python3-pip lsb-release
-$ sudo -H python3 -m pip install numpy scipy
+$ sudo apt install gdb lsb-release
```
Comment (SweetVishnya) [inline README.md:77]: rebase on master

File: casr/tests/tests.rs [hunk 4]
```diff
@@ -2430,20 +2430,19 @@ fn test_casr_cluster_c_huge_san() {
...
-            .count(),
-        1,
-        "Invalid number of reports in cluster 12"
-    );
+            .count();
+        if max < count {
+            max = count;
+        }
+    }
+    assert_eq!(max, 2, "Invalid max number of reports in cluster");
```
Comment (SweetVishnya) [inline casr/tests/tests.rs:2445]: We may get vector with cluster sizes, then sort it, and check it for expected values. The same below.

File: libcasr/src/stacktrace.rs (hunk 2 above)
Comment (SweetVishnya) [reply libcasr/src/stacktrace.rs]: Можно проверить, что все номера кластеров не нулевые

Comment (hkctkuy) [reply libcasr/src/stacktrace.rs]: Done

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

**Review by anfedotoff — APPROVED** [review]

**Comment by anfedotoff** [comment]
Before merging, we should check clustering on some real-world examples that we have from OSS-Sydr-Fuzz.

File: casr/tests/tests.rs (hunk 4 above)
Comment (hkctkuy) [reply casr/tests/tests.rs]: Done

**Review by SweetVishnya — APPROVED** [review]
Just two things left: 1) fix tests code style, 2) verify clusters for PyTorch.

**Pull request merged**